src/Pate/SimState.hs

```diff
@@ -75,49 +89,68 @@ import           GHC.Stack ( HasCallStack )
 import qualified Data.Kind as DK
 import           Data.Proxy
 
+import qualified Control.Monad.IO.Class as IO
 import           Control.Monad ( forM )
+import           Control.Lens ( (^.) )
+
+import qualified Prettyprinter as PP
+import           Prettyprinter ( (<+>) )
 
 import           Data.Parameterized.Classes
 import qualified Data.Parameterized.Context as Ctx
 import qualified Data.Parameterized.Map as MapF
-import           Data.Parameterized.Map ( Pair(..) )
 import qualified Data.Parameterized.TraversableF as TF
+import           Data.Functor.Const
 
 import qualified Data.Macaw.Symbolic as MS
 import qualified Data.Macaw.CFG as MM
 import qualified Data.Macaw.CFGSlice as MCS
+import qualified Data.Macaw.AbsDomain.AbsState as MAS
+import qualified Data.Macaw.Types as MT
 
 import qualified Lang.Crucible.LLVM.MemModel as CLM
 import qualified Lang.Crucible.Simulator as CS
 import qualified Lang.Crucible.Types as CT
 
 import qualified What4.Interface as W4
+import qualified What4.Concrete as W4
 import qualified What4.Expr.Builder as W4B
 
 import qualified Pate.Binary as PBi
 import qualified Pate.Block as PB
 import qualified Pate.ExprMappable as PEM
 import qualified Pate.Memory.MemTrace as MT
 import qualified Pate.PatchPair as PPa
+import qualified Pate.Panic as P
+import qualified Pate.Register.Traversal as PRt
 import qualified Pate.SimulatorRegisters as PSR
 import           What4.ExprHelpers
 import qualified Data.Parameterized.SetF as SetF
 
 ------------------------------------
 -- Crucible inputs and outputs
 
-
+-- | Points to the base of the stack. In any given context this will always be
+-- "free" as the base of the stack is always abstract, but it is rebound to account
+-- for changes to the stack pointer when moving between scopes.
+type StackBase sym arch v = ScopedExpr sym (W4.BaseBVType (MM.ArchAddrWidth arch)) v
```

Comment:
My usual annoying question: if we want to name this as something special, does it deserve a `newtype`?

Reply:
Agreed, I've wrapped it in a `newtype` now